flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -183,18 +214,40 @@ void notifySubpartitionConsumed() {
 	}
 
 	/**
-	 * Releases all received buffers and closes the partition request client.
+	 * Releases all received and available buffers, closes the partition request client.
 	 */
 	@Override
 	void releaseAllResources() throws IOException {
 		if (isReleased.compareAndSet(false, true)) {
+
+			final List<MemorySegment> recyclingSegments = new ArrayList<>();
+
 			synchronized (receivedBuffers) {
 				Buffer buffer;
 				while ((buffer = receivedBuffers.poll()) != null) {
-					buffer.recycle();
+					if (buffer.getRecycler() == this) {
+						recyclingSegments.add(buffer.getMemorySegment());
```

Comment:
I also considered this issue when implementation. The current way seems more verbose than directly calling `recycle()`, but there are two advantages:

1. Exclusive buffers are recycled in batch which may be more performant as you mentioned above. 
2. If calling `recycle()` directly, it should add extra check `isReleased` process outside the synchronized in `RemoteInputChannel#recycle` method.  It seems not elegant to do so. What do you think of this issue?

Reply:
1) I think, performance is not much of an issue here, as this is only called during take-down of a connection and the overhead of `Buffer#recycle` is actually not that much.
2) Sorry, but I don't get your concern. Why would you need an extra check when using `buffer.recycle()` instead of `exclusiveRecyclingSegments.add(buffer.getMemorySegment())`? There shouldn't be anything special for the exclusive buffers in this regard compared to ordinary buffers (which is the beauty of the design).

Let me give the example of how `LocalBufferPool` handles this inside `lazyDestroy`: it returns every memory segment (one by one) with `networkBufferPool.recycle()` and, at its end, it is calling `networkBufferPool.destroyBufferPool()` so that the book-keeping inside the `NetworkBufferPool` is updated and buffers may be re-distributed to other `LocalBufferPool` instances.
We could do this similarly here: recycle one by one, then call some method to update book-keeping and balancing inside `NetworkBufferPool`.

Reply:
oh, now I see - but let me elaborate on the `RemoteInputChannel#recycle()` there

Reply:
ok - regarding `RemoteInputChannel#recycle()`, if we use this here (indirectly via `buffer.recycle`), we would not want it to redistribute buffers for every single `recycle()` call, but if `RemoteInputChannel#recycle()` is called for a single buffer outside `RemoteInputChannel#releaseAllResources()`, we would want it to do that...

You're right that this would create some mess there and we should probably keep your current solution unless something better comes to mind.